scrapy/downloadermiddlewares/httpcompression.py

```diff
@@ -45,8 +51,9 @@ def process_response(self, request, response, spider):
                     # responsetypes guessing is reliable
                     kwargs['encoding'] = None
                 response = response.replace(**kwargs)
-                if not content_encoding:
-                    del response.headers['Content-Encoding']
+                if not self.keep_encoding_header:
```

Comment:
The problem I see with this solution is that responses could end up with a "Content-Encoding" header that does not fully match the actual encoding.
I think the ideal solution for this would be #187, but while that is not available, `Response.flags` is probably the less ugly way, as mentioned in https://github.com/scrapy/scrapy/issues/1988#issuecomment-219405456

Reply:
No worries I put it in just as you were commenting. I need to slow down a little.

So first with the from_crawler classmethod I overlooked, have it be `return cls(crawler.settings)` similar to Ajax. Then if I want to continue that path instead of passing HttpCompressionMiddleware anything it should get called with the from_crawler method. Which should be accessible since HttpComp test imports what ajax does, besides `from scrapy.utils.test import get_crawler`. 

Then to fix the possible mix match of the header value. Should I double check that the bytes passed are the same encoding as the value specified? That would need to be before its decoded. Or are you thinking to have the header value that is passed with the returned response object has the header encoding value be that of what it was decoded to?

Then response flags. Looks like they are passes when getting a Response object or through the replace method that takes keyword pairs. So then reworking the logic at the tail end to simply replace the encoding header value when it's being kept cause I don't see a way to tack on an additional header value.